standard/types.md

```diff
@@ -701,3 +701,4 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
 - `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `char`, `float`, `double`, `decimal`, or `bool`.
 - Any *enum_type*.
 - Any user-defined *struct_type* that is not a constructed type and contains instance fields of *unmanaged_type*s only.
+- In unsafe code ([§23.2](unsafe-code.md#232-unsafe-contexts)), any *pointer_type* ([§23.3](unsafe-code.md#233-pointer-types)).
```

Comment:
Suggested update (Assuming the type parameter must have the unmanaged constraint):

```suggestion
- Any user-defined *struct_type* that contains instance fields of *unmanaged_type*s or of type parameters with the `unmanaged` constraint only.
- In unsafe code ([§23.2](unsafe-code.md#232-unsafe-contexts)), any *pointer_type* ([§23.3](unsafe-code.md#233-pointer-types)).
```

Reply:
It is not clear that the `unmanaged` constraint is required, just that the provided type is unmanaged.

Have fun with the below:

- With the `where` clause the current VS set to 8.0
  - rejects the 3rd & 4th `CheckUnmanaged()` calls at compile time – they are not unmanaged types
  - accepts the second even though there is no `unmanaged` constraint on `Coords<>`
- However take out the `where` and it runs and reports only one of the types is managed, which is a little worrying! But then I copied the test code from StackOverflow…

```
using System;
using System.Collections.Generic;
using System.Reflection;

namespace UnmanagedConstructed
{
    public struct Coords<T>
    {
        public T X;
        public T Y;
    }

    public static class UnmanagedTypeExtensions
    {
        // U and IsUnmanaged copied from StackOverflow...
        class U<T> where T : unmanaged { }
        public static bool IsUnmanaged(this Type t)
        {
            try { typeof(U<>).MakeGenericType(t); return true; }
            catch (Exception) { return false; }
        }
    }

    public class UnmanagedTypes
    {
        public static void Main()
        {
            CheckUnmanaged<int>();
            CheckUnmanaged<Coords<int>>();
            CheckUnmanaged<object>();           // invalid if below where clause is uncommented
            CheckUnmanaged<Coords<object>>();   // ditto
        }

        private static void CheckUnmanaged<T>() // where T : unmanaged
        {
            Console.WriteLine($"{typeof(T)} is {(typeof(T).IsUnmanaged() ? "unmanaged" : "managed")}");
        }
    }
}
```

Reply:
After discussion, we agree with Nigel's analysis, and think we don't need to refer to type parameters, but we want an example (potentially like Nigel's, but simpler - just the method with an `unmanaged` constraint).

Reply:
@jskeet Is someone assigned to create the example? (I can't remember from the last meeting)

Reply:
@BillWagner: Me - the example along with the other changes to this PR. If I can just find the time...

Reply:
I've been having another look at this, and trying to get my head back into where we were in May... and I've failed. I'll mark this as "meeting discuss" again to see if we can get it right together *during* the next meeting.

I think there's some subtlety around type parameters though - because a type parameter with no constraints won't count as as unmanaged, but a type parameter with the unmanaged constraint will:

```csharp
class Program
{
    static void Main()
    {
        Check1<int>();
        Check2<int>();
    }

    static void Check1<T>()
    {
        CheckUnmanaged<T>();
    }

    static void Check2<T>() where T : unmanaged
    {
        CheckUnmanaged<T>();
    }

    static void CheckUnmanaged<T>() where T : unmanaged
    {
    }
}
```

Reply:
> However take out the where and it runs and reports only one of the types is managed, which is a little worrying! But then I copied the test code from StackOverflow…

We believe that `Coords<object>` being reported as unmanaged is just a runtime bug - it really *is* a managed type, and so this is incorrect. But out of the scope of the spec.

Reply:
The `IsUnmanaged` runtime check using `MakeGenericType` in the sample above does not match the compiler's behavior, and the runtime decided this was by design (https://github.com/dotnet/runtime/issues/12420#issuecomment-479983685) citing the original 7.3 proposal, which is quoted there as saying:

> The unmanaged constraint is not enforced by CLR, only by the language.

Reply:
Ah. Closing the loop on this, the reason that one fails and not the other has nothing to do with `unmanaged`. It's just the runtime checking the implicit `struct` constraint that gets emitted when `unmanaged` is used. The runtime totally ignores `unmanaged` as a concept.